programmazione-funzionale.md

```diff
@@ -1 +1,198 @@
+
+```java
+Stream<String> stream = persone.stream()
+    .filter(p -> p.getEta() >= 30)
+    .map(p -> {
+        return p.getNome() + ": " + 
+        p.getEta().toString();
+    });
+    /* return implicito 
+    (valido solo per funzioni 
+    con una sola riga) */
+```
+
+Otterrò uno `stream`, contenente le stringhe: "Luigi: 30", "Pippo: 40", "Pluto: 50".
+
+### `Distinct` e `Sort`
+
+- **`distinct()`** elimina gli elementi duplicati dello `stream`.
+
+- **`sort()`** ordina gli elementi dello `stream`. Ha bisogno di un `Comparator` per oggetti diversi da `Integer` e `String`.
+
+---
+
+### `flatMap`
+
+**`flatMap(<funzione da T a Stream<U>>)`** prende in input una funzione, la applica ad ogni elemento di tipo T dello `stream`. La funzione ritorna un altro `stream` (in generale contentente elementi di tipo diverso U) per ogni elemento. Infine tutti gli `stream` vengono concatenati in un unico `stream`.
+
+Nel nostro esempio, vogliamo trasformare ogni persona (sempre filtrata con età >= 30) in una lista di stringhe contenenti il nome e l'età, e poi concatenare tutti gli `stream` in un unico `stream`:
+
+```java
+Stream<String> stream = persone.stream()
+    .filter(p -> p.getEta() >= 30)
+    .flatMap(p -> {
+        List<String> listaPerP;
+        listaPerP = new ArrayList<>();
+        listaPerP.add(p.getNome());
+        listaPerP.add(p.getEta().toString());
+        return listaPerP.stream();
+    });
+```
+
+Otterrò uno `stream`, contenente le stringhe: "Luigi", "30", "Pippo", "40", "Pluto", "50".
+
+### `Reduce`
+
+La **`reduce(Identità, <funzione binaria>)`** è un'operazione per avere un singolo valore a partire da uno `stream`. 
+
+Si ha come primo parametro un valore iniziale (identità), e come secondo parametro una funzione binaria che prende in input due elementi dello `stream` e ritorna un altro elemento (dello stesso tipo) ottenuto a partire dai due. La funzione verrà applicata ricorsivamente ad ogni coppia di elementi dello `stream`, fino ad ottenere un solo elemento.
+
+Nel nostro esempio, vogliamo ottenere la somma delle età delle persone (filtrate):
+
+```java
+Integer sommaEta = persone.stream()
+    .filter(p -> p.getEta() >= 30)
+    /* mapToInt è uguale alla map, 
+    ma ritorna uno stream di interi */
+    .mapToInt(p -> p.getEta())  
+    .reduce(
+        0, 
+        (eta1, eta2) -> eta1 + eta2 );
+```
+
+Otterrò un intero: 120 (= 30 + 40 + 50).
+
+### `Collect`
+
+Siamo ora interessati ad avere una Collezione (List, Set, ...) invece che uno stream. Per fare ciò, possiamo usare il metodo **`collect(<funzione che ritorna una Collezione>)`**. Nel nostro esempio, vogliamo ottenere una lista di stringhe contenente nome: età delle persone (filtrate con età >= 30):
+
+```java
+List<String> lista = persone.stream()
+    .filter(p -> p.getEta() >= 30)
+    .map(p -> 
+        p.getNome() + ": " + 
+        p.getEta().toString()
+    );
+    .collect(Collectors.toList());
```

Comment:
Se riesci spiega la faccenda dei collectors, di come si implementa per una collezione custom e del fatto che per le liste è già definito. (è una cosa secondaria, si può fare anche più in la nelle prossime iterazioni di scrittura).